rfd/0094-discover-metrics.md

```diff
@@ -0,0 +1,195 @@
+* _Where do users experience failures when adding resources?_ For example,
+  agent does not join in time or connection tester fails.
+
+Most of the questions we're interested in can be answered using PostHog
+[funnels](https://posthog.com/manual/funnels) as they are designed to help
+understand how users progress through the product flows and where the
+bottlenecks and friction are.
+
+To be able to build a Discover funnel, each step the user goes through needs
+to be defined as an event. Below is a Protobuf spec containing proposed events
+metadata. Note that each event will include anonymized cluster name and timestamp
+as shown in our [reporting spec](https://github.com/gravitational/prehog/blob/main/proto/prehog/v1alpha/teleport.proto)
+so below we only include event metadata.
+
+```protobuf
+// Contains common metadata for Discover related events.
+message DiscoverMetadata {
+    // Uniquely identifies Discover wizard "session". Will allow to correlate
+    // events within the same Discover wizard run. Can be UUIDv4.
+    string id = 1;
+}
+
+// Contains common metadata identifying resource type being added.
+message ResourceMetadata {
+    // Resource type: server, kubernetes, database, app, desktop.
+    string type = 1;
+    // Optional protocol where applicable, e.g. postgres, mysql, sqlserver.
+    string protocol = 2;
+}
+
+// Contains fields that track a particular step outcome, for example connection
+// test failed or succeeded, or user aborted the step.
+message Status {
+    // Indicates the step outcome. For example, "success" means user proceeded
+    // to the next step, "error" if something failed (e.g. connection test),
+    // "aborted" if user exited the wizard.
+    string status = 1;
+    // Contains error details. We have to be careful to not include any
+    // identifyable infomation like server addresses here.
+    string error = 2 ;
+}
+
+// Emitted when user selected resource type to add and proceeded to the next
+// step.
+message DiscoverResourceSelectionEvent {
+    DiscoverMetadata discover_metadata = 1;
+    ResourceMetadata resource_metadata = 2;
+    Status status = 3;
+}
+
+// Emitted on the "Configure Resource" screen when user has connected an SSH
+// or Kubernetes agent and proceeded to the next step.
+//
+// For Database Access this step has sub-steps so will have individual events
+// defined below, however this event will be emitted as well at the end of
+// the last sub-step to be able to track it on a more high-level.
+message DiscoverConfigureResourceEvent {
+    DiscoverMetadata discover_metadata = 1;
+    ResourceMetadata resource_metadata = 2;
+    Status status = 3;
+}
+
+// Emitted when a user registered a database resource and proceeded to the next
+// step.
+message DiscoverRegisterDatabase {
+    DiscoverMetadata discover_metadata = 1;
+    ResourceMetadata resource_metadata = 2;
+    Status status = 3;
+}
+
+// Emitted when a user deployed a database agent and proceeded to the next step.
```

Comment:
We have another step here: setting the labels and generating a token (happens in the same screen)
Should we add an event for that as well?

Reply:
I thought about the same but wasn't sure if that actually provides a lot of value for now. TBH I think we could've just started with "start" and "stop" event and that already would be better than what we have now, but eventually I've settled for "1 event per screen" which should give us enough granularity to understand how users flow through the wizard.